src/librbd/Operations.cc

```diff
@@ -1510,7 +1497,14 @@ int Operations<I>::prepare_image_update() {
   }
   m_image_ctx.owner_lock.get_read();
 
-  return r;
+  if (r < 0) {
+    return r;
+  } else if (m_image_ctx.exclusive_lock != nullptr &&
+             !m_image_ctx.exclusive_lock->is_lock_owner()) {
+    return -EROFS;
```

Comment:
@dillaman I saw this many times and used myself but was alway wondering if this is entirely correct. I suppose there is a chance that in a time window between we acquired the exclusive lock and got the owner lock the acquired lock was stolen again by other client request. We will return -EROFS in this case though might handle this situation a better way (e.g. block acquire lock requests from other clients until we get the owner lock)?

Also I am not sure I understand the `accept_requests()` check above. It looks like the purpose is to block local operations until it is true, but I don't see why `accept_requests()` is expected to be always true after `try_acquire_lock()` is completed?

Reply:
It could use a loop to ensure it keeps the lock. With regard to ```accept_request```, it looks like it's a carry over from this tracker ticket [1] for ensuring that the journal has been fully initialized before allowing the operation to proceed.

[1] http://tracker.ceph.com/issues/14510

Reply:
Ah, so `accept_requests()` is used here for `is_state_locked()` test and not for block_requests as I thought initially.

As for using loop I suppose it is a bit troublesome to distinguish between the case when the lock was stolen and the case when another client did not want to release it? We want to loop only in the first case. Why not just increase `m_request_blocked_count` before calling `try_acquire_lock()` (putting the owner lock) and decreasing it when getting the owner lock back?

Reply:
Sure -- tweak pushed